### PR TITLE
List View: Allow right-click to open block settings dropdown, add editor setting

### LINKED ISSUE #40287: Hijack the browser's right-click contextual menu when right-clicking on any part of the items in the FSE / Gutenberg post's blocks hierarchy list view sidebar
## What problem does this address?

With the list view sidebar activated on the left side of the Gutenberg editor, whether for posts/pages or for the FSE, I'm always annoyed when I right-click those items in the list and end up with my browser's (Firefox's) contextual right-click menu full of options irrelevant to my intended action regarding that interactive non-content widget (the list view), which was to get the item-specific actions that Gutenberg offers in its "three dots" menu for that item.

Currently I have the intuitive habit of right-clicking the items anywhere, getting the useless browser menu, having to left-click to dismiss it, then having to left-click the three dots to get the correct menu. I shouldn't have to aim for the tiny three-dots menubutton to get that menu, I should be able to right-click anywhere on that item's row. Fitt's law, etc., for a more efficient and native-feeling/integrated application UX.

## What is your proposed solution?

Please hijack the browser's right-click event to show the three-dots contextual menu of Gutenberg actions when clicking on a block item, particularly in the list view sidebar. If Google Drive and GMail can do it, so can WordPress.

Comparison of the current¹ vs suggested behavior when right-clicking anywhere on a row/item in the list view:

wordpress gutenberg FSE suggestion to hijack right-click menus

¹: Tested with the stable versions bundled with WordPress 5.9.3 and the TwentyTwentyTwo theme, on Firefox 99 on Linux.

### PR DESCRIPTION
**Pull request by andrewserong**
## What?

Fixes #40287, based on a previous attempt in #41041

This PR adds right-click behaviour to the list view, so that right-clicking on a list view item will open the block settings menu, with the menu positioned over the mouse cursor.

It also adds an editor setting to the post and site editors so that users can disable the setting if they wish.

## Why?

The general idea, as raised in https://github.com/WordPress/gutenberg/issues/40287, is that in other web apps that people are used to, right-clicking an item opens up a contextual menu that is part of the app. The more that using Gutenberg feels like using an app, the more it might be appropriate that right-click feels like more of an in-app kind of behaviour. That's the idea, at least!

## How?

* Add an `allowRightClickOverrides` setting in the post and site editors.
* At the button level in the List View, add an `onContextMenu` handler, and if the settings allow it, simulate a click on the settings dropdown when the button is right-clicked.

## Testing Instructions

1. In the post editor, add some blocks.
2. Open up the list view, and right click on an item in the list. The block settings menu should open up.
3. Multi-select multiple blocks and right click on an item — you should still be able to remove multiple blocks via this dropdown.
4. Open up the Preferences in the editor, and toggle the right click overrides setting off.
5. Reload the editor for the change in preferences to take effect.
6. Check that the browser default context menu opens up when right clicking list view items.
7. Repeat the above in the site editor.

## Screenshots or screencast 

Right-click items in the list view:

https://github.com/WordPress/gutenberg/assets/14988353/f0a62f23-36d9-48df-982e-b477a5390212

The setting:

| Post editor setting | Site editor setting |
| --- | --- |
| image | image |

### CODE REVIEW

**Comment by andrewserong** [comment]
> Sometimes, when right-clicking, the other menus don't disappear:

I believe I can reliably reproduce this in the following situation:

* Have focus outside of the browser window, e.g. in another application like screen sharing software
* Mouse over the list view but do a right click before doing anything else
* Notice that the settings menu opens up, but focus isn't shifted to the first item in the list
* Right click on another list view item — the first popover will not be closed, even though the next one opens up

This appears to be because focus was never in the first popover (or quite possibly within the window at all), so opening another popover doesn't cause the focus outside logic to fire and close the original popover.

I'm not _too_ sure how best to fix this just yet, but one possible idea could be to keep track of an `id` of the settings menu that's currently open, and when it changes, then close all the other ones? That feels a little state-heavy to me as an idea, so there could be a better way to do it 🤔

**Comment by andrewserong** [comment]
Update: you can replicate the same thing via CMD-click and left clicking on the settings dropdown buttons, so I _think_ there could be an argument for fixing this outside of the right-click use case, too

**Comment by andrewserong** [comment]
Update: in https://github.com/WordPress/gutenberg/pull/50273/commits/61ef62af39f0e237d2de6a7a059d94aaf7ffc704 I've fixed up the initial flicker of the list view row being focused before opening up the settings menu.

The last blocker appears to be the issue of the block settings menu not closing when focus isn't on the browser window. I'll continue digging into that.

**Comment by andrewserong** [comment]
I've written up an issue for the problem of multiple block settings menus being displayed over in: https://github.com/WordPress/gutenberg/issues/50988

**Comment by mtias** [comment]
Thanks for picking this one up again! It will be even better when we combine with the new flyout menu component so we can reduce how long the initial state is and group things a bit better

**Comment by andrewserong** [comment]
> It will be even better when we combine with the new flyout menu component so we can reduce how long the initial state is and group things a bit better

Absolutely! And from the sounds of it, the new `DropdownMenu` component will likely resolve the outstanding issue in this PR (right clicking other items in the list view sometimes doesn't cause the original dropdown menu to close). My thinking at the moment is to wait for the new component to be ready before attempting to merge this PR, but do let me know if anyone would like to see this right click option land sooner.

**Comment by andrewserong** [comment]
Update: it looks like the refactor to using the v2 of `DropdownMenu` has encountered some blockers (over in https://github.com/WordPress/gutenberg/pull/51400). If it doesn't make it in time for 6.3, then I would recommend we park this PR also for 6.3 due to the bug where multiple dropdown menus sometimes appear / don't close properly. I think it's probably a better idea to ensure the dropdown menu is behaving smoothly before exposing it as the primary UI for right-click within the list view.

Do let me know if anyone feels differently, though, and I can give this PR a rebase / get it as close to ready as it can be for now.

**Comment by andrewserong** [comment]
Update: in https://github.com/WordPress/gutenberg/pull/50273/commits/130c0d726b81f4f8956121a454db0b9535054608, I've had a go at introducing some state in the block editor to track the clientId of the block settings menu that is currently open. If an id is present and it doesn't match a particular block settings menu, then we can force it to close. This seems to fix the issue:

https://github.com/WordPress/gutenberg/assets/14988353/f305b5d7-c2b1-4f49-b646-2f123409178c

That fix is a little beyond the scope of the overall intention behind this PR (introduce right-click support), so I'll open up separate PR(s) for that fix to see if it winds up looking viable, and will link over in https://github.com/WordPress/gutenberg/issues/50988. Once that issue is resolved in `trunk`, we can then rebase this PR and get it ready for review.

**Comment by andrewserong** [comment]
Update: now that https://github.com/WordPress/gutenberg/pull/54083 has landed, I've given this a rebase and it seems to mostly be working properly now. I'm not sure if we've 100% eradicated issues involving the block settings menu opening and closing correctly, so it'd be worth giving this a manual test to see how it feels for folks in actual usage.

**Review by apeatling — COMMENTED** [review]
This seems to be working well for me, I could not seem to break it and cause multiple menus to show. It works in both the site and post editor (although didn't work when editing a page -- I think by design).

Not shown in the gif is multi block select which I also confirmed to work correctly.

**Comment by ramonjd** [comment]
Tested in FF, Safari, Chrome and Edge

- Block settings contextual menu appears on right click over list view items (but not image previews or the ellipsis menus)
- Multiple selected items have the correct, single context menu on right click
- Menu functionality works (I couldn't test paste styles in FF because I didn't have https running)

https://github.com/WordPress/gutenberg/assets/6458278/4336b0e3-cc67-4b35-8be1-7cc60fddec07

Pretty impressive!

When you come back to this PR, would it be worth creating a small E2E test that runs through [a right mouse click](https://playwright.dev/docs/input#mouse-click) to ensure the context menu appears?

Other than that I'd say it's good to go unless you're waiting on further UX feedback ?

File: packages/block-editor/src/components/list-view/block.js
Comment (ramonjd) [inline packages/block-editor/src/components/list-view/block.js:237]: Is it that `ownerDocument` can never be falsy due to the fallback value of `{}`?

Comment (andrewserong) [reply packages/block-editor/src/components/list-view/block.js:237]: This one's slightly subtle — `ownerDocument` is set via destructuring, so if the fallback value after the assignment operator is `{}` then `ownerDocument` will be `undefined` as it doesn't exist on `{}` 🙂

File: packages/edit-site/src/components/preferences-modal/index.js
Comment (ramonjd) [inline packages/edit-site/src/components/preferences-modal/index.js:69]: To me, `enableListViewRightClick` or `allowListViewContextualMenus` or the equivalent specific descriptor might describe the feature. `allowRightClickOverrides` might be too generic?

File: packages/edit-site/src/components/preferences-modal/index.js
Comment (ramonjd) [inline packages/edit-site/src/components/preferences-modal/index.js:72]: Do we need to communicate that it's specific to the List view here? I don't think it has effect anywhere else in the editor (?)

File: packages/block-editor/src/components/list-view/block.js
Comment (ramonjd) [reply packages/block-editor/src/components/list-view/block.js:237]: Oh right, sorry I read that line wrong. `{}` is the fallback for `rowRef?.current` 🤦🏻

File: packages/edit-site/src/components/preferences-modal/index.js
Comment (andrewserong) [reply packages/edit-site/src/components/preferences-modal/index.js:69]: Good question! My thinking was to keep it generic so that if we wish to expand other areas to also have right click overrides, we don't wind up with a proliferation of `allowRightClick` settings in the preferences 🤔

File: packages/edit-site/src/components/preferences-modal/index.js
Comment (andrewserong) [reply packages/edit-site/src/components/preferences-modal/index.js:72]: Oh, for the user facing help text, that's a great idea! Perhaps `Allows contextual list view menus via right-click...` or something like that.

**Comment by andrewserong** [comment]
Thanks for reviewing @ramonjd! I haven't had any design or product feedback on this feature for 6.4, and given the tight timing before feature freeze for 6.4, I'm thinking it might be best to park this one until 6.5? My main thinking here is that while I think this feature is working well, I'd prefer not to introduce it prematurely as it's a fairly prominent change if folks are used to using the right mouse button in that area.

> When you come back to this PR, would it be worth creating a small E2E test that runs through [a right mouse click](https://playwright.dev/docs/input#mouse-click) to ensure the context menu appears?

Yes, good idea! I'll add that in 👍

**Comment by ramonjd** [comment]
> I haven't had any design or product feedback on this feature for 6.4, and given the tight timing before feature freeze for 6.4, I'm thinking it might be best to park this one until 6.5? My main thinking here is that while I think this feature is working well, I'd prefer not to introduce it prematurely as it's a fairly prominent change if folks are used to using the right mouse button in that area.

Sounds very reasonable! Could we move it to Punted to 6.5 col and let it stew in the plugin for a while?

**Comment by andrewserong** [comment]
> Could we move it to Punted to 6.5 col and let it stew in the plugin for a while?

Sounds good! Done 👍

Thanks for the nudging, and yes, it'll be much better to try this feature out in the plugin and see what people think of it before introducing it to core 🙂

**Comment by mittyan** [comment]
```
Add an allowRightClickOverrides setting in the post and site editors.
At the button level in the List View, add an onContextMenu handler, and if the settings allow it, simulate a click on the settings dropdown when the button is right-clicked.
```

The worst instructions I ever read after the "draw an owl meme". 
Can someone point me how and where can I enable these options? Theme.json, functions.php, where? Which hooks/functions should I use? Dude..

**Comment by andrewserong** [comment]
> Can someone point me how and where can I enable these options?

The feature doesn't yet exist in Gutenberg/WordPress, but this PR proposes introducing it. The description of this PR describes how the feature was implemented. It is not something that's configurable from within themes.

For folks who haven't checked out the Gutenberg source code, the feature can be tested in the post editor via the following URL: http://gutenberg.run/50273

**Comment by mtias** [comment]
After #56481 lands, let's keep this setting in general in both editors.

**Comment by andrewserong** [comment]
> After https://github.com/WordPress/gutenberg/pull/56481 lands, let's keep this setting in general in both editors.

Thank you, will do!

**Comment by andrewserong** [comment]
Now that #56481 has landed, I've given this PR a rebase and retested that it still appears to be working well. So, I think this should be ready for a final review! 🤞

The goal with this PR is that we're adding in right-click in the list view, to open the block settings menu for the currently hovered list view item. In follow-ups we could look at expanding right click behaviour into the editor canvas, and see what might make for good behaviour there, too.

File: packages/edit-post/src/components/preferences-modal/index.js
Comment (ramonjd) [inline packages/edit-post/src/components/preferences-modal/index.js]: Do you think we should specify that the contextual menus display in the list view?

**Review by ramonjd — APPROVED** [review]
Checked in site and post editors:

✅  With the preferences turned on the list view contextual menu appears on right click. All menu items function as expected.
✅  Turned off, I see the browser contextual menu only.

Works in FF, Safari, Chrome and Edge. 

Great stuff!!

I left a comment about being specific about where the menu shows, maybe to set expectations? 😄

File: packages/edit-post/src/components/preferences-modal/index.js
Comment (andrewserong) [reply packages/edit-post/src/components/preferences-modal/index.js]: At the moment, I was thinking of leaving it as-is, because of the potential to add right-click behaviour to the editor canvas in follow-ups, where we'll re-use the setting. Happy to change it to specify list view for now, if it's confusing, though!

**Comment by andrewserong** [comment]
Thanks for reviewing @ramonjd! 🙇

It looks like the e2e failures are legit — it's [an unfortunate e2e test](https://github.com/WordPress/gutenberg/blob/c1852f305c58e08979cc8d6093d9ecfd57a5585a/test/e2e/specs/editor/various/a11y.spec.js#L123) that passes for me locally but not in the Github Action because our Github Actions use a different window size to my local environment, and the test expected that the "General" tab in preferences wasn't long enough to cause a scrollbar. I think that test likely needs to be rewritten a little (or at least to have comments added) as it's highly likely to break when adding or removing preferences from the panel. I'll see if I can fix it up for this PR, though.

In terms of functionality in this PR, nothing's breaking, it's just that the test is failing 🙂

**Comment by ramonjd** [comment]
I kicked them off again, but I think maybe the [failing e2e tests in test/e2e/specs/editor/various/a11y.spec.js](https://github.com/WordPress/gutenberg/actions/runs/7203713429/job/19625276269?pr=50273#step:7:191) are related because the preferences pane has been updated?

**Comment by ramonjd** [comment]
Sorry, my page hadn't updated by the time you wrote the [comment](https://github.com/WordPress/gutenberg/pull/50273#issuecomment-1855086522)!

> I think that test likely needs to be rewritten a little (or at least to have comments added) as it's highly likely to break when adding or removing preferences from the panel. I'll see if I can fix it up for this PR, though.

Makes sense, thanks for explaining.

**Comment by andrewserong** [comment]
I've pushed a temporary fix for the e2e test in https://github.com/WordPress/gutenberg/pull/50273/commits/35d42be5a8a05746ea9a340948c1044dccf59fa8 to to hopefully resolve the issue for this PR (assuming the actions pass now!). Separately, it could be good to come up with another way of testing that particular part of the preferences panel as it should ideally test the behaviour but not be dependent on the current content of the preferences panels 🤔

File: packages/edit-post/src/components/preferences-modal/index.js
Comment (andrewserong) [reply packages/edit-post/src/components/preferences-modal/index.js]: Updated in https://github.com/WordPress/gutenberg/pull/50273/commits/452376760e91e594a877428b9de1a1b30e7c9e56 to include `list view`:

image

**Comment by ramonjd** [comment]
> Separately, it could be good to come up with another way of testing that particular part of the preferences panel as it should ideally test the behaviour but not be dependent on the current content of the preferences panels

Yeah, that seems like a flakey approach. What you've seems like a good compromise, while still preserving the test. 

Thanks!

LGTM

**Review by apeatling — APPROVED** [review]
Gave this another solid test and I couldn't break it, looking good! 👍

**Comment by andrewserong** [comment]
Thanks for the reviews, folks! I'll merge this in and then open up a follow-up issue for ideas for other areas that could benefit from a custom right-click dropdown.

**Comment by andrewserong** [comment]
Follow-up issue for discussing other places where right-click could work: https://github.com/WordPress/gutenberg/issues/57091

**Comment by jameskoster** [comment]
Love this!

**Comment by richtabor** [comment]
Same! Good stuff!

**Pull request merged**
